### DISCUSSION TITLE
MySQL supported ?

### DISCUSSION THREAD

**Discussion by DevXiaolan** [FerretDB/FerretDB#50] [General]
Is it in planning ?

[a comment by an automated account is not shown]

**Comment by muety** [comment]
I'd be interested!

**Reply by rhuddleston** [reply]
same here

**Reply by ptrfarkas** [reply]
Thanks for sharing Ryan, long time no talk! Out of curiosity, what would be the main reason for you to prefer MySQL over PostgreSQL? Just wanted to understand the specific reasons so we can prioritize this accordingly.  Your feedback would help us a lot.

**Reply by rhuddleston** [reply]
Hi Peter! Mostly just because we have a lot of operational expertise with MySQL and have HA clusters where we never take them down for any maintenance using group replication with near instant failover etc. We don't have these capabilities currently with postgres or Mongo.

**Reply by ptrfarkas** [reply]
Understood, thanks Ryan!

**Reply by muety** [reply]
It's similar for me. Mainly, I just have a lot more experience working with and operating MySQL than Postgres. I'm using MySQL for almost all software that I use, including my own applications as well as third-party software, so if FerretDB supported MySQL, that'd fit in well.

**Reply by muety** [reply]
Anything in planning for MySQL support?

**Reply by ptrfarkas** [reply]
@muety Thank you for your continued interest, we really appreciate it! When we started implementing FerretDB, we decided to focus on a single database backend to ensure strong support, rather than providing weak support for multiple backends. We chose Postgres due to its robust community support and JSON compatibility, which simplified the development process. We are still working to achieve a high degree of compatibility, so our primary focus remains on Postgres. That being said, we are taking measures to facilitate contributions for additional database backends, such as SQLite and MySQL. Our goal is to incorporate support for these databases as soon as possible, but these efforts will only scale appropriately if we get some contributors to help out with each.  We'll release more info on how to contribute soon.

In the meanwhile, any details on your specific use case helps us with understanding what features would be needed in order to make this a viable alternative for you.

**Reply by Conan-Kudo** [reply]
I'd like to mention that the reason I'd prefer MariaDB/MySQL over PostgreSQL is the lower operational complexity. My environment isn't very big, and MariaDB is fairly easy to handle in that environment. In particular, database software upgrades with MariaDB are much less complex than with PostgreSQL because MariaDB can handle in-place upgrades while PostgreSQL cannot. There are ugly workarounds for doing this documented by the PostgreSQL community, and some Linux distributions have implemented their own versions of these schemes. 

Hopefully this provides a concrete example of the operational burden of PostgreSQL for smaller environments that leads to a preference for MariaDB/MySQL.

**Reply by ptrfarkas** [reply]
@Conan-Kudo thank you for your valuable feedback.

As an update on this, we are getting closer (but, unfortunately not close) to supporting MySQL as we are in the process of rolling out our new architecture. This will make it much easier to add new database backends to FerretDB. 

We are hoping that the new architecture we are rolling out soon will make it possible to get MySQL on our roadmap. We will be creating GitHub issues for the missing features, with the expectation that community contributors will be picking them up. 

In the meantime our support for SQLite will be production ready in our next release, which could be used in less complex environments.

**Comment by AlekSi** [comment]
#3413 outlines the process of how MySQL backend could be added.
